torch/_decomp/decompositions.py

```diff
@@ -1225,6 +1225,83 @@ def cudnn_batch_norm_backward(
     )
 
 
+@register_decomposition(aten._adaptive_avg_pool2d)
+def adaptive_avg_pool2d(input: Tensor, output_size: Tuple[int, int]):
+    # Preconditions
+    device = input.device
+    shape = input.shape
+    ndim = len(shape)
+    utils.check(
+        ndim in (3, 4),
+        lambda: f"adaptive_avg_pool2d(): Expected 3D or 4D tensor, but got {ndim}",
+    )
+    for d in input.shape[-2:]:
+        utils.check(
+            d != 0,
+            lambda: "adaptive_avg_pool2d(): Expected input to have non-zero size for "
+            f"non-batch dimensions, but input has shape {tuple(shape)}.",
+        )
+
+    # Optimisation (we should also do this in the kernel implementation)
+    if shape[-2] % output_size[-2] == 0 and shape[-1] % output_size[-1] == 0:
+        stride = tuple(i // o for i, o in zip(shape[-2:], output_size))
+        kernel = tuple(
+            i - (o - 1) * s for i, o, s in zip(shape[-2:], output_size, stride)
+        )
+        return torch.nn.functional.avg_pool2d(input, kernel, stride)
+
+    def start_index(a, b, c):
+        return (a * c).float().div(b).floor().long()
+
+    def end_index(a, b, c):
+        return ((a + 1) * c).float().div(b).ceil().long()
+
+    # Let's assume the reduction we want to apply is to sum all the elements (averaging from this is easy)
+    # Even more, let's assume that we want to just do the 1d case.
+    # The 2d case is recovered by applying the 1d case along two dimensions
+    # The issue here is that we may want to sum segments of different sizes.
+    # What we do is to get the largest segment, and select all the elements from the initial points
+    # up to the max length. Then we zero out the elements that we picked up and were not necessary
+
+    def adaptive_avg_pool1d(x, dim, out_size):
+        assert dim == -2 or dim == -1
+        in_size = x.dim(dim)
+
+        orange = torch.arange(out_size, device=device)
+        i0 = start_index(orange, out_size, in_size)
+        i1 = end_index(orange, out_size, in_size)
+        length = i1 - i0
+
+        # Select elements
+        # Doing this logic with arange is intrinsically data-dependent, as the length of the range
+        # is data-dependent. I don't think there's a way to do this that's not data-dependent really
+        maxlength = int(torch.max(length).item())
+        ones = torch.ones(maxlength - 1, dtype=i0.dtype, device=device)
+        idx = torch.cat([i0.unsqueeze(1), ones]).cumsum(dim=1)  # [size, max]
```

Comment:
torch.cat will result in horrible performance in backends, don't use it in decomps

Reply:
It is not clear to me how to implement the batched version of `arange` otherwise. What do you propose?

Reply:
Could we keep these as separate tensors and not concat them?

Reply:
ooops. That is certainly the right way of doing this!